### ISSUE TITLE
tempdir is still used

### ISSUE THREAD

**Issue by pverbaan** [simplesamlphp/simplesamlphp#1960]
**Specifics of your environment**
1. Are you acting as SP/IdP/proxy? IdP
2. SimpleSAMLphp: What version are you using? 2.1.2
3. PHP: What version are you using? 8.0.30
4. Platform: unix or Windows? unix
5. Webserver: Apache/Nginx/ISS? Apache + Nginx

**Describe the bug**
An UNHANDLEDEXCEPTION is thrown when loading the admin page.

**To Reproduce**
1. Go to '/simplesaml/module.php/admin/'
2. See error

**Expected behavior**
The login page to appear

**Screenshots or logs**
```
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] SimpleSAML\Error\Error: UNHANDLEDEXCEPTION
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] Backtrace:
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 1 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/src/SimpleSAML/Error/ExceptionHandler.php:32 (SimpleSAML\Error\ExceptionHandler::customExceptionHandler)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 0 [builtin] (N/A)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] Caused by: SimpleSAML\Assert\AssertionFailedException: /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/config/config.php: Could not retrieve the required option 'tempdir'.
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] Backtrace:
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 9 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/vendor/simplesamlphp/assert/src/Assert.php:364 (SimpleSAML\Assert\Assert::__callStatic)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 8 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/src/SimpleSAML/Configuration.php:401 (SimpleSAML\Configuration::getValue)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 7 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/src/SimpleSAML/Configuration.php:690 (SimpleSAML\Configuration::getString)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 6 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/src/SimpleSAML/Kernel.php:57 (SimpleSAML\Kernel::getCacheDir)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 5 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/vendor/symfony/http-kernel/Kernel.php:346 (Symfony\Component\HttpKernel\Kernel::getBuildDir)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 4 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/vendor/symfony/http-kernel/Kernel.php:437 (Symfony\Component\HttpKernel\Kernel::initializeContainer)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 3 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/vendor/symfony/http-kernel/Kernel.php:767 (Symfony\Component\HttpKernel\Kernel::preBoot)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 2 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/vendor/symfony/http-kernel/Kernel.php:190 (Symfony\Component\HttpKernel\Kernel::handle)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 1 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/src/SimpleSAML/Module.php:234 (SimpleSAML\Module::process)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] 0 /var/www/vhosts/<snipped>/simplesamlphp-2.1.2/public/module.php:17 (N/A)
[Mon Feb 12 09:42:39 2024] simplesamlphp ERROR [d96575e64a] Error report with id 9c0748dd generated.
```

**Additional context**
The `tempdir` setting was removed in the default configuration in #1933, but the settings is still used (as a fallback) in `SimpleSAML\Kernel::getCacheDir`:
```
        $cachePath = $configuration->getOptionalString(
            'cachedir',
            $configuration->getString('tempdir')
        ) . DIRECTORY_SEPARATOR . $this->module;
```
The workaround is to uncomment the `tempdir` setting in the configuration.

This seems like a low priority issue, and I reported it more as a heads-up for other people that might encounter the same thing.

**Comment by thijskh** [comment]
I understand that the message is confusing, but the code is correct. The template indeed removed `tempdir` but added `cachedir`. You need either one, if both are missing, then you get this error.

You could improve the code to throw a more concrete exception when you miss both. But in essence this fix is to have a `cachedir` in your `config.php`.

**Comment by pverbaan** [comment]
Ah, it seems you missed my point. I have a `cachedir` setting in the configuration, but the code attempts to resolve `tempdir` first (in `$configuration->getString('tempdir')`, before it returns the `cachedir` setting. Unfortunately, the code is not executed lazily.

So, even though the `tempdir` setting is not actually used, it is being resolved.

**Comment by tvdijen** [comment]
I can see where I went wrong... I propose a fix like this:

```
$temp = $configuration->getOptionalString('tempdir');
$cache = $configuration->getOptionalString('cachedir');
$cacheDir = $cache ?? $temp;
Assert::notNull($cacheDir, "Must configure a cachedir.");

$cachePath = $cacheDir . DIRECTORY_SEPARATOR . $this->module;
```

**Comment by thijskh** [comment]
Ah yes, indeed that looks better (maybe change string to `Missing cachedir parameter in config.php`

**Comment by tvdijen** [comment]
Closed in 6f1203a7f
